### PR TITLE
Add OpenTelemetry support for Koa template

### PR DESCRIPTION
**Pull request by runk**
This brings https://opentelemetry.io/ support, which can be enabled via gantry config (deploys another sidecar that collects the data). According to @bernos it's here to stay and should no longer be considered an experimental feature. Yet has to be widely announced. 

All traces can be viewed in "Technology Platforms" DD org.

This PR is a first step - one template only - koa app.

### CODE REVIEW

File: template/koa-rest-api/src/tracing.ts
Comment (runk) [inline template/koa-rest-api/src/tracing.ts:36]: I'm hesitant to bring in the standard logger as it might bring more dependencies before internals instrimented. This lib really meant to be self-contained.

Thoughts?

File: template/koa-rest-api/src/listen.ts
Comment (runk) [inline template/koa-rest-api/src/listen.ts:1]: It's loaded in app.ts

File: template/koa-rest-api/gantry.apply.yml
Comment (runk) [inline template/koa-rest-api/gantry.apply.yml:27]: There's a convention currently with this naming. Assuming it'd be great to keep it that way.

File: template/koa-rest-api/package.json
Comment (samchungy) [inline template/koa-rest-api/package.json:10]: question: do you need to declare this as a dep?

File: template/koa-rest-api/package.json
Comment (samchungy) [inline template/koa-rest-api/package.json:6]: question: should we be using ^ for the versions for consistency?

**Comment by etaoins** [comment]
Has this been running in production yet? Any early performance insights versus no instrumentation or Datadog APM?

File: template/koa-rest-api/package.json
Comment (runk) [reply template/koa-rest-api/package.json:10]: I think it's up for discussion. It's needed for `aws-sdk` instrumentation, which I guess is highly desirable for anyone at seek. Perhaps some don't use `aws-sdk` but I bet it's the minority.

Comment (samchungy) [reply template/koa-rest-api/package.json:10]: I think when I flipped through the GitHub page it said you could use aws-sdk V3 as well?

**Comment by runk** [comment]
> Has this been running in production yet? Any early performance insights versus no instrumentation or Datadog APM?

Otel is in production for about ~20 services today, however only some have code instrumentation enabled. Most of those 20 services have the flag set in gantry which enables tracing in auth-sidecar *only*. I can see there are only 4 services that have code instrumentation enabled across a couple of teams. There're some heavy-loaded systems like the API router that is instrumented though. Also it comes down that what we want to instrument, e.g. both `aws-sdk` and node's `http` or only one one of those.

Re DD APM - I think it'd be a similar performance hit as with automatic instrumentation as all outgoing requests (and other i/o commands) are wrapped. 

Stats for one of the services:

Screen Shot 2021-12-01 at 9 22 04 am

Screen Shot 2021-12-01 at 9 19 09 am

I'll try to get similar insights for another service that's hit a bit harder.

**Comment by runk** [comment]
Okay, there're two discreet parts - app code instrumentation (controlled via `tracing.ts`) and auth sidecar instrumentation (controlled via gantry config). The previous chart was for the case when we already had Otel enabled in gantry (think auth sidecar) and just added code instrumentation in the app itself. No visible performance hit could be seen.

However, when enabled in gantry AND app, there's about 2-3ms latency increase. I'm inclined to believe it's mainly attributed to sidecar instrumentation.


Screen Shot 2021-12-01 at 9 53 15 am

**Comment by etaoins** [comment]
Is that expected @bernos? I don't recall that showing up on `indie-graphql-server`. Seems like a large regression for a feature teams might not end up using.

**Review by etaoins — APPROVED** [review]
Talking with @runk offline, there's potentially a large performance regression for some workloads (33%-50%). I'm happy to include this as a default off to allow teams to opt-in to tracing if they find it useful. This is similar to what we do with Datadog APM.

**Comment by runk** [comment]
Changed to be opt-in.

I think as an alternative (a middle ground?) we can to have the tracing on in auth sidecar by default, but keep app instrumentation off by default. I think this way possible performance hit as you experienced with DD APM will be an unlikely scenario.

**Comment by runk** [comment]
Screen Shot 2021-12-01 at 11 10 07 am

The graph shows the app in three modes:

- Bare app (before 9:15)
- Sidecar + App instrumentation (9:15-10:38)
- Sidecar instrumentation only (10:38 onwards)

While latency increase is a single digit, it could mean a ~30% increase for a fast-responding app. Assuming if the app is doing a lot of network calls and all be instrumented (e.g. Experience APIs/Graphql servers), the more serious performance hit of app instrumentation should be expected. I do not see time spent for sidecar instrumentation to differ depending on nature of the app itself.

File: template/koa-rest-api/package.json
Comment (bernos) [reply template/koa-rest-api/package.json:10]: Actually, otel won't work with aws-sdk < V3. At least it wasn't in early testing. This was because in V2 of the aws SDK, request signing was happening _before_ otel header propagation. This meant that all instrumented requests to AWS services via the AWS SDK would fail on account of invalid signatures. This wasn't an issue in V3

File: template/koa-rest-api/src/tracing.ts
Comment (bernos) [inline template/koa-rest-api/src/tracing.ts:15]: Does `async` code loaded via `--require` behave any differently to `async` code called from the entry point? ie. is there a race here between instrumentation being initialised and the first requests being handed by the service?

**Comment by bernos** [comment]
This is great, happy to roll it out, but please don't default any open telemetry options to `enabled` by default just yet. The cost implications of mass adoption are still not well understood, and we'd like to avoid any shock there. There are contingencies and compensations that we can make in the event that costs exceed the forecast, but we need a bit more control over the rate of adoption in order to have time to adapt if needs be. Also, I mentioned an incompatibility with aws-sdk < v3 that may have been addressed in newer releases of v2, but there should probably be a warning on the tin.

File: template/koa-rest-api/package.json
Comment (runk) [reply template/koa-rest-api/package.json:10]: @bernos re `aws-sdk` v2 - I can confirm instrumentation works for it with current otel libraries.

Comment (samchungy) [reply template/koa-rest-api/package.json:10]: Would it be better to just leave it out and let consumers byo sdk?

File: template/koa-rest-api/src/tracing.ts
Comment (runk) [reply template/koa-rest-api/src/tracing.ts:15]: Good point. I think it's definitely a possibility that promises aren't fulfilled before the app loads. Today there's no way to pause for async execution in the bootstrap file supplied via `--require x.js`.

Really depends on what happens inside of the `sdk.start` fn: https://www.npmjs.com/package/@opentelemetry/sdk-node. Looking in code it only really returns a pending promise when we want auto resource detection which is off: https://github.com/seek-oss/skuba/pull/706/files#diff-4836bf38ef9493029c23deb46fdf06733bbc8b8cec520b9f538ea1a68cebf8eaR33

I'm going to remove async from the fn to keep it simple. However it's only hiding the possible problem with pending promise from `sdk.start()`

File: template/koa-rest-api/package.json
Comment (runk) [reply template/koa-rest-api/package.json:10]: It depends on your definition of "better", as always :). Do we want to bring in `aws-sdk` instrumentation for consumers out of the box? 

Screen Shot 2021-12-01 at 4 21 38 pm

Looking at this trace - I see clear advantages of knowing what the HTTP POST was actually is about, which in the case with `aws-sdk` instrumentation is very illustrative. With minimal built-in `http` instrumentation it'll be just another opaque HTTP request.

The problem is https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-sdk needs aws-sdk (or rather its types)

**Comment by runk** [comment]
@samchungy @etaoins do you want to check it one more time before I merge? The notable change is the addition of process exit handling: https://github.com/seek-oss/skuba/pull/706/files#diff-4836bf38ef9493029c23deb46fdf06733bbc8b8cec520b9f538ea1a68cebf8eaR46

**Comment by samchungy** [comment]
> @samchungy @etaoins do you want to check it one more time before I merge? The notable change is the addition of process exit handling: https://github.com/seek-oss/skuba/pull/706/files#diff-4836bf38ef9493029c23deb46fdf06733bbc8b8cec520b9f538ea1a68cebf8eaR46

@72636c your review probably holds the most weight here 😆

File: template/koa-rest-api/src/listen.ts
Comment (72636c) [reply template/koa-rest-api/src/listen.ts:1]: I think this is here to make it obvious (to humans but also to e.g. linters) that the import has side effects and must go first. It's easy to shoot yourself in the foot by importing something else before `./app`, which is part of the reason why I'd like to get rid of this hook soon™️.

File: template/koa-rest-api/package.json
Comment (72636c) [inline template/koa-rest-api/package.json:8]: Do we need to be concerned about these v0s? Does OpenTelemetry promise some degree of stability?

File: template/koa-rest-api/Dockerfile
Comment (72636c) [inline template/koa-rest-api/Dockerfile:34]:
```suggestion
CMD ["--require", "lib/tracing.js", "lib/listen.js"]
```

Or the other way; I'm not fussed as long as we're consistent:

```suggestion
CMD ["--require", "./lib/tracing.js", "./lib/listen.js"]
```

File: .changeset/brown-bananas-repair.md
Comment (72636c) [inline .changeset/brown-bananas-repair.md:5]:
```suggestion
template/koa-rest-api: Add opt-in OpenTelemetry support
```

Q: should we add this to the Express template too?

File: template/koa-rest-api/src/tracing.ts
Comment (72636c) [inline template/koa-rest-api/src/tracing.ts:2]:
```suggestion
 * OpenTelemetry tracing initialisation. This is a standalone TS/JS module that is not
```

File: template/koa-rest-api/src/tracing.ts
Comment (72636c) [inline template/koa-rest-api/src/tracing.ts:41]:
```suggestion
    .then(() => log('info', 'OpenTelemetry initialised'))
```

File: template/koa-rest-api/gantry.apply.yml
Comment (72636c) [inline template/koa-rest-api/gantry.apply.yml:15]:
```suggestion
  OPENTELEMETRY_ENABLED: '{{.Values.openTelemetry.enabled | default false}}'
  SERVICE: '{{values "service"}}'
```

File: template/koa-rest-api/src/tracing.ts
Comment (72636c) [inline template/koa-rest-api/src/tracing.ts:23]:
```suggestion
  // Use B3 propagation to ensure proper propagation between systems that use
  // OpenTelemetry and native Datadog APM, such as Istio/Envoy.
```

File: .changeset/brown-bananas-repair.md
Comment (runk) [reply .changeset/brown-bananas-repair.md:5]: Sure, do you want me to do it in the same PR or separate?

File: template/koa-rest-api/src/tracing.ts
Comment (72636c) [inline template/koa-rest-api/src/tracing.ts:15]: For posterity, I was wondering where this `app` log field convention was from and it looks like our sidecar follows suit.

**Review by 72636c — APPROVED** [review]

File: .changeset/brown-bananas-repair.md
Comment (72636c) [reply .changeset/brown-bananas-repair.md]: Feel free to merge this one in first if that's easier

File: template/koa-rest-api/src/tracing.ts
Comment (etaoins) [reply template/koa-rest-api/src/tracing.ts:15]: Old mate `ca-go-library` 🙈

`seek-auth-sidecar` was originally `ca-auth-sidecar` and used the default logger from `ca-go-library`. The newer `name` convention comes from Pino/Bunyan.

**Pull request merged**
